woocommerce/woocommerce
# Add variations autoselect

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:




TL;DR: Added options to auto-select variation attributes that only have one possible value.
- Added a new settings tab: Interface
- Added the functionality to auto-select attributes according to other selected attributes (both per-block and globally)
- Added the functionality to auto-select attributes right after page loads (both per-block and globally)
- Added the functionality to display attributes in conflict and select what to do with them (both per-block and globally)
- The initial options are now filtered once after page load, according to possible variations
- Replaced `option_gt_filter` by `[value=""]` for readability, simplicity, and avoid corner cases
- All defaults mirror current behaviors



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### Prepare a new test product

1. Create a new product with the following settings:
    1. `Product type`: `Variable product`
    2. `Attributes`:
        1. `"Type"`: `"T-Shirt"`
        2. `"Colour"`: `"Red|Blue"`
        3. `"Size"`: `"L|XL"`
        3. Click on `Save attributes`
    3. `Variations`:
        1. `Type`: `"T-Shirt"`, `Colour`: `"Red"`, `Size`: `"L"`
        2. `Type`: `"T-Shirt"`, `Colour`: `"Red"`, `Size`: `"XL"`
        3. `Type`: `"T-Shirt"`, `Colour`: `"Blue"`, `Size`: `"XL"`
    4. Enable each variations and set a price
    5. Click on `Save variations`
2. Save and publish the new product

#### Configuring new settings

1. Navigate to `<YOUR_SITE_URL>/wp-admin/admin.php?page=wc-settings&tab=interface`
    (`Woocommerce > settings > Interface`)
2. Under `Attribute dropdown options`:
    1. Set `Auto-select behavior` to `All attributes`
    2. Check `"Auto-select on page load"`
    3. Set `Values in conflict with current selection` to `Grayed out and disabled`

#### Testing auto-select on page load behavior

1. Navigate the test product
2. Refresh the tab
3. The attributes should now be auto-selected (on page load) to:
    1. `Type`: `"T-Shirt"` (this is because `Type` has only 1 possible value (`"T-Shirt"`))
    2. `Colour`: `"Choose an option"` (`""`)
    3. `Size`: `"Choose an option"` (`""`)

#### Testing auto-select behavior

1. Navigate to the test product
2. Refresh the tab
3. The attributes should now be auto-selected (on page load) to:
    1. `Type`: `"T-Shirt"`
4. Select `Colour` > `"Blue"`
5. The remaining options will now be auto-selected according to the value set in `Configuring new settings` > `step #2.1`
6. In this case, `Size` will be auto-selected to `"XL"`

#### Testing the conflicting attribute values behavior

1. Navigate to the test product
2. Refresh the tab
3. Select `Colour` > `"Red"`,  `Size` > `"L"`
4. Click on the `Colour` attribute
5. Here is what will happen to the `"Blue"` option in the `Colour` dropdown, because it is conflicting with other attributes, according to the value set in `Configuring new settings` > `step #2.1`
    1. `Hidden`: The `"Blue"` option is not there, it was hidden from view (temporarily removed from DOM) (default action and behavior in the current version of woocommerce)
    1. `Grayed-out and disabled`: The `"Blue"` option is still there, but disabled and grayed-out (the option itself is disabled)
    1. `Grayed-out but selectable (will clear all other attribute if selected)`: The `"Blue"` option is still there, but grayed-out (the option itself is still enabled, but clicking it will produce an impossile variation, making all attributes reset)



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

Added auto-select for attribute fields; Possibility to Display attributes in conflict




</details>
